### DISCUSSION TITLE
FastAPI Example seems overcomplex

### DISCUSSION THREAD

**Discussion by HomerusJa** [agronholm/apscheduler#1000] [General]
I am trying to use APScheduler in my FastAPI project. When I looked at the example, I was a bit shocked on how seemingly complex the code seemed to be. Is there any specific reason to directly integrate with the ASGI instead of using FastAPI specific features? Following something I would think would also work (not tested as I am not at my PC right now, and man it is pain to write code on an iPhone):

```python

from datetime import datetime
from fastapi import FastAPI
from fastapi.responses import PlainTextResponse
from sqlalchemy.ext.asyncio import create_async_engine
from apscheduler import AsyncScheduler
from apscheduler.datastores.sqlalchemy import SQLAlchemyDataStore
from apscheduler.eventbrokers.asyncpg import AsyncpgEventBroker
from apscheduler.triggers.interval import IntervalTrigger

def tick():
    print("Hello, the time is", datetime.now())

engine = create_async_engine("postgresql+asyncpg://postgres:secret@localhost/testdb")
data_store = SQLAlchemyDataStore(engine)
event_broker = AsyncpgEventBroker.from_async_sqla_engine(engine)
scheduler = AsyncScheduler(data_store, event_broker)

app = FastAPI()

@app.on_event("startup")
async def start_scheduler():
    await scheduler.add_schedule(tick, IntervalTrigger(seconds=1), id="tick")
    await scheduler.start_in_background()

# Maybe also add a shutdown event

@app.get("/")
async def root():
    return PlainTextResponse("Hello, world!")

```

**Comment by HomerusJa** [comment]
Just saw this is discussion 1000 in this repo. Congrats 👏🏼

**Comment by agronholm** [comment]
`AsyncScheduler` must be used as an async context manager due to structured concurrency (`TaskGroup`). Also integrating with ASGI makes the example easier to adapt to other frameworks.

**Comment by HomerusJa** [comment]
I see the point that integrating directly with ASGI makes the example easier to adapt. However, I think in an example tailored towards a specific framework, you should try to integrate with that framework as usual, because in that case the example would be far easier to understand for people just programming with FastAPI which have never bothered with the ASGI internals. I might be wrong here, but I think that is most.

I wasnt aware you had to use an async context manager, but now I looked into the code (and learned about TaskGroups along the way) and understand why that is necessary. Still, you can use FastAPI's built in features here:

```python
# Do the basic scheduler setup here
scheduler = AsyncScheduler(data_store, event_broker)


@asynccontextmanager
async def lifespan(app: FastAPI):
    async with scheduler:
        await scheduler.add_schedule(tick, IntervalTrigger(seconds=1))
        await scheduler.start_in_background()
        yield

app = FastAPI(lifespan=lifespan)
```

**Reply by agronholm** [reply]
Alright, I'll keep this in mind, thanks.
